File: scripts/_faq_normalize.py
"""Shared normalization helpers for FAQPage JSON-LD validators.

Python-side mirror of src/lib/faq-normalize.ts. Every Python validator
(scripts/validate-library-faq.py, scripts/report-library-faq-jsonld.py)
routes duplicate + trimmed checks through here so JS and Python agree.

Keep the rule (whitespace collapse + strip + lower) identical to the TS
helper. If you change one, change both.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Callable, Iterable, Sequence, TypeVar

T = TypeVar("T")

_WHITESPACE_RUN = re.compile(r"\s+")


def normalize_faq_text(value: object) -> str:
    """Collapse internal whitespace, strip ends, lowercase."""
    if value is None:
        return ""
    return _WHITESPACE_RUN.sub(" ", str(value).strip()).lower()


def is_trimmed(value: object) -> bool:
    """True when value is a str with no leading/trailing whitespace."""
    return isinstance(value, str) and value == value.strip()


def has_normalized_content(value: object) -> bool:
    """True when normalized text has any visible characters."""
    return len(normalize_faq_text(value)) > 0


def is_nonempty_str(value: object) -> bool:
    """True when value is a str with visible content (post-strip)."""
    return isinstance(value, str) and len(value.strip()) > 0


def equals_normalized(a: object, b: object) -> bool:
    """True when both values are equal after shared FAQ normalization.

    Mirrors src/lib/faq-normalize.ts `equalsNormalized`. Use for every
    FAQ JSON-LD string comparison (@type, @context, name, text) so
    trivial whitespace/case differences do not cause spurious failures.
    """
    return normalize_faq_text(a) == normalize_faq_text(b)


def type_matches_normalized(type_field: object, expected: str) -> bool:
    """JSON-LD @type may be str OR list[str]; match under normalization."""
    values = type_field if isinstance(type_field, list) else [type_field]
    return any(equals_normalized(t, expected) for t in values)


def context_matches_schema_org(context_field: object) -> bool:
    """True when any @context entry references schema.org after normalization."""
    values = context_field if isinstance(context_field, list) else [context_field]
    for c in values:
        if not isinstance(c, str):
            continue
        if normalize_faq_text(c).rstrip("/").endswith("schema.org"):
            return True
    return False


@dataclass(frozen=True)
class DuplicateGroup:
    key: str
    indices: tuple[int, ...]


def find_duplicate_groups(
    items: Sequence[T],
    extractor: Callable[[T, int], object],
) -> list[DuplicateGroup]:
    """Group items by normalize_faq_text(extractor(...)); keep groups of >=2."""
    buckets: dict[str, list[int]] = {}
    for i, item in enumerate(items):
        key = normalize_faq_text(extractor(item, i))
        buckets.setdefault(key, []).append(i)
    return [
        DuplicateGroup(key=k, indices=tuple(idxs))
        for k, idxs in buckets.items()
        if len(idxs) > 1
    ]


def format_duplicate_groups(
    groups: Iterable[DuplicateGroup], key_label: str = "value"
) -> list[str]:
    return [
        f'"{g.key[:80]}" at mainEntity[{",".join(str(i) for i in g.indices)}] ({key_label})'
        for g in groups
    ]

File: src/lib/faq-normalize.ts
// Shared normalization helpers for FAQPage JSON-LD validators.
//
// The live sweep (library-faq-jsonld.test.ts), the negative regression suite
// (library-faq-jsonld-negative.test.ts), and any future JS/TS validator MUST
// route through these helpers so "duplicate" and "trimmed" checks agree
// everywhere. If you change the rule, change it here — not inline.

/** Collapse internal whitespace, trim ends, lowercase. Used for equality. */
export function normalizeFaqText(input: unknown): string {
  return String(input ?? "")
    .replace(/\s+/g, " ")
    .trim()
    .toLowerCase();
}

/** True when input is a string with no leading/trailing whitespace. */
export function isTrimmed(input: unknown): boolean {
  return typeof input === "string" && input === input.trim();
}

/** True when input has visible content after whitespace normalization. */
export function hasNormalizedContent(input: unknown): boolean {
  return normalizeFaqText(input).length > 0;
}

/**
 * True when two values compare equal after shared FAQ normalization
 * (whitespace collapse + trim + lowercase). Use for any FAQ JSON-LD
 * string field comparison (@type, @context, name, text, etc.) so tests
 * fail only on real semantic differences, not trivial formatting.
 */
export function equalsNormalized(a: unknown, b: unknown): boolean {
  return normalizeFaqText(a) === normalizeFaqText(b);
}

/**
 * JSON-LD `@type` can be a string OR an array of strings. Returns true
 * when any type entry matches `expected` under equalsNormalized.
 */
export function typeMatchesNormalized(typeField: unknown, expected: string): boolean {
  const list = Array.isArray(typeField) ? typeField : [typeField];
  return list.some((t) => equalsNormalized(t, expected));
}

/**
 * True when any @context entry (string or array) references schema.org
 * after normalization. Accepts http/https, trailing slashes, and mixed case.
 */
export function contextMatchesSchemaOrg(contextField: unknown): boolean {
  const list = Array.isArray(contextField) ? contextField : [contextField];
  return list.some((c) => {
    if (typeof c !== "string") return false;
    return normalizeFaqText(c).replace(/\/+$/, "").endsWith("schema.org");
  });
}

export interface DuplicateGroup {
  /** Normalized key that collided. */
  key: string;
  /** Indices in the original array that share this normalized key. */
  indices: number[];
}

/**
 * Group items by normalizeFaqText and return only groups with >=2 members.
 * Extractor picks the string field to compare (e.g. `e => e.name`).
 */
export function findDuplicateGroups<T>(
  items: readonly T[],
  extractor: (item: T, index: number) => unknown,
): DuplicateGroup[] {
  const buckets = new Map<string, number[]>();
  items.forEach((item, i) => {
    const key = normalizeFaqText(extractor(item, i));
    const bucket = buckets.get(key);
    if (bucket) bucket.push(i);
    else buckets.set(key, [i]);
  });
  const dupes: DuplicateGroup[] = [];
  for (const [key, indices] of buckets) {
    if (indices.length > 1) dupes.push({ key, indices });
  }
  return dupes;
}

/** Format a DuplicateGroup[] for test / report error messages. */
export function formatDuplicateGroups(
  groups: readonly DuplicateGroup[],
  keyLabel = "value",
): string[] {
  return groups.map(
    (g) => `"${g.key.slice(0, 80)}" at mainEntity[${g.indices.join(",")}] (${keyLabel})`,
  );
}
